components/hts221/hts221.c

```diff
@@ -0,0 +1,397 @@
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_set_data_rate(hts221_t *dev, hts221_data_rate_t dr)
+{
+    CHECK_ARG(dev && dr <= HTS221_12_5HZ);
+
+    return update_reg(dev, REG_CTRL_REG1, dr, MASK_CTRL_REG1_ODR);
+}
+
+esp_err_t hts221_get_heater(hts221_t *dev, bool *enable)
+{
+    CHECK_ARG(dev && enable);
+
+    uint8_t raw;
+    CHECK(read_reg(dev, REG_CTRL_REG2, &raw));
+    *enable = (raw & BIT(BIT_CTRL_REG2_HEATER)) ? true : false;
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_set_heater(hts221_t *dev, bool enable)
+{
+    CHECK_ARG(dev);
+
+    return update_reg(dev, REG_CTRL_REG2, enable ? BIT(BIT_CTRL_REG2_HEATER) : 0, BIT(BIT_CTRL_REG2_HEATER));
+}
+
+esp_err_t hts221_get_averaging(hts221_t *dev, hts221_temperature_avg_t *t_avg, hts221_humidity_avg_t *rh_avg)
+{
+    CHECK_ARG(dev && (t_avg || rh_avg));
+
+    uint8_t raw;
+    CHECK(read_reg(dev, REG_AV_CONF, &raw));
+    if (t_avg)
+        *t_avg = (raw >> BIT_AV_CONF_AVGT0) & MASK_AV_CONF_AVG;
+    if (rh_avg)
+        *t_avg = (raw >> BIT_AV_CONF_AVGH0) & MASK_AV_CONF_AVG;
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_set_averaging(hts221_t *dev, hts221_temperature_avg_t t_avg, hts221_humidity_avg_t rh_avg)
+{
+    CHECK_ARG(dev && t_avg <= HTS221_AVGT_256 && rh_avg <= HTS221_AVGH_512);
+
+    return write_reg(dev, REG_AV_CONF, (t_avg << BIT_AV_CONF_AVGT0) | (rh_avg << BIT_AV_CONF_AVGH0));
+}
+
+esp_err_t hts221_get_drdy_config(hts221_t *dev, bool *enable, hts221_drdy_mode_t *mode, hts221_drdy_level_t *active)
+{
+    CHECK_ARG(dev && (enable || mode || active));
+
+    uint8_t raw;
+    CHECK(read_reg(dev, REG_CTRL_REG3, &raw));
+
+    if (enable)
+        *enable = (raw >> BIT_CTRL_REG3_DRDY) & 1;
+    if (mode)
+        *mode = (raw >> BIT_CTRL_REG3_PP_OD) & 1;
+    if (active)
+        *active = (raw >> BIT_CTRL_REG3_DRDY_H_L) & 1;
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_set_drdy_config(hts221_t *dev, bool enable, hts221_drdy_mode_t mode, hts221_drdy_level_t active)
+{
+    CHECK_ARG(dev && mode <= HTS221_DRDY_OPEN_DRAIN && active <= HTS221_DRDY_ACTIVE_LOW);
+
+    return write_reg(dev, REG_CTRL_REG3,
+            (enable ? BIT(BIT_CTRL_REG3_DRDY) : 0) | (mode << BIT_CTRL_REG3_PP_OD) | (active << BIT_CTRL_REG3_DRDY_H_L));
+}
+
+esp_err_t hts221_is_data_ready(hts221_t *dev, bool *ready)
+{
+    CHECK_ARG(dev && ready);
+
+    uint8_t raw;
+    CHECK(read_reg(dev, REG_STATUS_REG, &raw));
+
+    *ready = (raw & MASK_STATUS_REG_DA) == MASK_STATUS_REG_DA;
+
+    return ESP_OK;
+}
+
+esp_err_t hts221_start_oneshot(hts221_t *dev)
+{
+    CHECK_ARG(dev);
+
+    return update_reg(dev, REG_CTRL_REG2, BIT(BIT_CTRL_REG2_ONE_SHOT), BIT(BIT_CTRL_REG2_ONE_SHOT));
+}
+
+esp_err_t hts221_get_data(hts221_t *dev, float *t, float *rh)
+{
+    CHECK_ARG(dev && (t || rh));
+
+    int16_t raw_rh;
+    int16_t raw_t;
+    uint8_t raw[4];
+
+    I2C_DEV_TAKE_MUTEX(&dev->i2c_dev);
+    I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_HUMIDITY_OUT_L, raw, 4));
```

Comment:
I have tested it, and the sensor returns incorrect values. 

When I replace 
``` 
I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_HUMIDITY_OUT_L, raw, 4)); 
```
with 
```
I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_HUMIDITY_OUT_L, &raw[0], 1));
I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_HUMIDITY_OUT_H, &raw[1], 1));
I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_TEMP_OUT_L, &raw[2], 1));
I2C_DEV_CHECK(&dev->i2c_dev, i2c_dev_read_reg(&dev->i2c_dev, REG_TEMP_OUT_H, &raw[3], 1));
````
the sensor returns correct temperature and humidity values.